src/backend/jack/mod.rs

```diff
@@ -250,33 +261,21 @@ impl MidiOutput {
         handler_data
     }
     
-    pub fn connect(mut self, port_number: usize, port_name: &str) -> Result<MidiOutputConnection, ConnectError<MidiOutput>> {
-        let dest_port_name = {
-            let ports = self.client.as_ref().unwrap().get_midi_ports(PortIsInput);
-            if port_number >= ports.count() {
-                None
-            } else {
-                Some(ports.get_c_name(port_number).to_owned()) // have to copy the name to prevent borrowing issues
-            }
-        };
-        
-        let dest_port_name = match dest_port_name {
-            None => return Err(ConnectError::new(ConnectErrorKind::PortNumberOutOfRange, self)),
-            Some(s) => s
-        };
-        
+    pub fn connect(mut self, port: &MidiOutputPort, port_name: &str) -> Result<MidiOutputConnection, ConnectError<MidiOutput>> {
         let mut handler_data = self.activate_callback();
         
         // Create port ...
-        let port = match self.client.as_mut().unwrap().register_midi_port(port_name, PortIsOutput) {
+        let source_port = match self.client.as_mut().unwrap().register_midi_port(port_name, PortIsOutput) {
```

Comment:
The use of `unwrap` is disallowed in real life code. In this specific case, even when the function is returning a `Result`, there might be a panic, which would be breaking the contract of the function. I suggest you chain errors, or else use `if let`.

Reply:
Yep, that's usually true, but in this case, we know that `self.client` will actually never be `None` except between the `self.client.take()` in line 282 and the end of the method. And if it actually was `None` here, it's clearly a programming error on my side, and a panic is totally the right thing to happen then (it's a little bit like an implied `assert!(self.client.is_some())`).

It would be much better without using `Option` here, but I didn't find a way to do that when I originally wrote that code. I'll have to revisit this and see if there's a better way.

Reply:
Thanks for the explanation, I've been quite involved in functional programming lately, and couldn't avoid to see that, but your argument is completely reasonable :-P
FYI While looking for possible alternatives I learnt about the `do-notation` in rust [here](https://m4rw3r.github.io/rust-questionmark-operator), but I haven't used it myself yet.